File: Core/Source/NSScanner+HTML.m
//
//  NSScanner+HTML.m
//  DTCoreText
//
//  Created by Oliver Drobnik on 1/12/11.
//  Copyright 2011 Drobnik.com. All rights reserved.
//

#import "NSScanner+HTML.h"
#import "NSCharacterSet+HTML.h"
#import "DTColorFunctions.h"
#import "NSString+HTML.h" 

@implementation NSScanner (HTML)

#pragma mark CSS

// scan a single element from a style list
- (BOOL)scanCSSAttribute:(NSString * __autoreleasing*)name value:(id __autoreleasing*)value
{
	NSString *attrName = nil;
	
	NSInteger initialScanLocation = [self scanLocation];
	
	NSCharacterSet *whiteCharacterSet = [NSCharacterSet whitespaceAndNewlineCharacterSet];
	
	NSMutableCharacterSet *nonWhiteCharacterSet = [[NSCharacterSet whitespaceAndNewlineCharacterSet] mutableCopy];
	[nonWhiteCharacterSet formUnionWithCharacterSet:[NSCharacterSet characterSetWithCharactersInString:@";"]];
	[nonWhiteCharacterSet invert];

	NSMutableCharacterSet *nonWhiteCommaCharacterSet = [[NSCharacterSet whitespaceAndNewlineCharacterSet] mutableCopy];
	[nonWhiteCommaCharacterSet formUnionWithCharacterSet:[NSCharacterSet characterSetWithCharactersInString:@";,"]];
	[nonWhiteCommaCharacterSet invert];

	
	// alphanumeric plus -
	NSCharacterSet *cssStyleAttributeNameCharacterSet = [NSCharacterSet cssStyleAttributeNameCharacterSet];
	
	if (![self scanCharactersFromSet:cssStyleAttributeNameCharacterSet intoString:&attrName])
	{
		return NO;
	}
	
	// skip whitespace
	[self scanCharactersFromSet:whiteCharacterSet intoString:NULL];
	
	// expect :
	if (![self  scanString:@":" intoString:NULL])
	{
		[self setScanLocation:initialScanLocation];
		return NO;
	}
	
	// skip whitespace
	[self scanCharactersFromSet:whiteCharacterSet intoString:NULL];
	
	NSMutableArray *results = [NSMutableArray array];
	BOOL nextIterationAddsNewEntry = YES;
	
	while (![self isAtEnd] && ![self scanString:@";" intoString:NULL])
	{
		// skip whitespace
		[self scanCharactersFromSet:whiteCharacterSet intoString:NULL];

		NSString *quote = nil;
		if ([self scanCharactersFromSet:[NSCharacterSet quoteCharacterSet] intoString:&quote])
		{
			NSString *quotedValue = nil;
			
			// attribute is quoted
			if (![self scanUpToString:quote intoString:&quotedValue])
			{
				[self setScanLocation:initialScanLocation];
				return NO;
			}
			else
			{
				if (nextIterationAddsNewEntry)
				{
					[results addObject:quotedValue];
					nextIterationAddsNewEntry = NO;
				}
				else
				{
					quotedValue = [NSString stringWithFormat:@"%@ %@%@%@", [results lastObject], quote, quotedValue, quote];
					[results removeLastObject];
					[results addObject:quotedValue];
				}
			}
			
			// skip ending quote
			[self scanString:quote intoString:NULL];
			
			//TODO: decode unicode sequences like "\2022"
		}
		else
		{
			// attribute is not quoted, we append elements until we find a ; or the string is at the end
			NSString *valueString = nil;
			
			if ([self scanString:@"rgb(" intoString:&valueString])
			{
				if ([valueString isEqualToString:@"rgb("])
				{
					[self scanUpToString:@";" intoString:&valueString];
					NSString * formattedRGBString = [NSString stringWithFormat:@"rgb(%@", valueString];
					
					if (nextIterationAddsNewEntry)
					{
						[results addObject:formattedRGBString];
						nextIterationAddsNewEntry = NO;
					}
					else
					{
						valueString = [NSString stringWithFormat:@"%@ %@", [results lastObject], formattedRGBString];
						[results removeLastObject];
						[results addObject:valueString];
					}
				}
			}
			else if ([self scanString:@"," intoString:&valueString])
			{
                BOOL isStringOnlyCSSProperty = NO;
                
				if (![valueString isEqualToString:@","])
				{
					[results addObject:valueString];
				}
				else if ([attrName isEqualToString:@"font"] ||
						 ([attrName rangeOfString:@"color"].location != NSNotFound) ||
						 ([attrName rangeOfString:@"shadow"].location != NSNotFound) ||
						 ([attrName rangeOfString:@"background"].location != NSNotFound))
				{
					valueString = [NSString stringWithFormat:@"%@%@", [results lastObject], valueString];
					[results removeLastObject];
					[results addObject:valueString];
                    
                    isStringOnlyCSSProperty = YES;
				}
				
				if ([valueString isEqualToString:@","] && !isStringOnlyCSSProperty)
				{
					nextIterationAddsNewEntry = YES;
				}
			}
			else if ([self scanCharactersFromSet:nonWhiteCommaCharacterSet intoString:&valueString])
			{
				if ([valueString length] && ![valueString isEqualToString:@","])
				{
					if (nextIterationAddsNewEntry) {
						[results addObject:valueString];
						nextIterationAddsNewEntry = NO;
					} else {
						valueString = [NSString stringWithFormat:@"%@ %@", [results lastObject], valueString];
						[results removeLastObject];
						[results addObject:valueString];
					}
				}
			}
		}

		// skip whitespace
		[self scanCharactersFromSet:whiteCharacterSet intoString:NULL];
	}
	
	// Success 
	if (name)
	{
		*name = attrName;
	}
	
	if (value)
	{
		if (results.count == 0) {
			*value = @"";
		} else if (results.count == 1) {
			*value = [results objectAtIndex:0];
		} else {
			*value = results;
		}
	}
	
	return YES;
}

/*
 
 Source: http://www.w3.org/TR/CSS1/#url
 
 The format of a URL value is 'url(' followed by optional white space followed by an optional single quote (') or double quote (") character followed by the URL itself (as defined in [11]) followed by an optional single quote (') or double quote (") character followed by optional whitespace followed by ')'. Quote characters that are not part of the URL itself must be balanced.
 
 Parentheses, commas, whitespace characters, single quotes (') and double quotes (") appearing in a URL must be escaped with a backslash: '\(', '\)', '\,'.
 
 Partial URLs are interpreted relative to the source of the style sheet, not relative to the document:
*/

// NOTE: Simplified, we assume that there are no quotes in the URL

- (BOOL)scanCSSURL:(NSString * __autoreleasing*)urlString
{
	if (![self scanString:@"url(" intoString:NULL])
	{
		return NO;
	}
	

	NSCharacterSet *quoteCharacterSet = [NSCharacterSet quoteCharacterSet];
	NSString *quote;
	NSString *attrValue;
	
	if ([self scanCharactersFromSet:quoteCharacterSet intoString:&quote])
	{
		if ([quote length]==1)
		{
			[self scanUpToString:quote intoString:&attrValue];	
			[self scanString:quote intoString:NULL];
		}
		else
		{
			// most likely e.g. href=""
			attrValue = @"";
		}
		
		// decode HTML entities
		attrValue = [attrValue stringByReplacingHTMLEntities];
	}
	else 
	{
		// non-quoted attribute, ends at )
		if ([self scanUpToString:@")" intoString:&attrValue])
		{
			// decode HTML entities
			attrValue = [attrValue stringByReplacingHTMLEntities];
		}
	}

	if (urlString)
	{
		*urlString = attrValue;
	}
	
	return YES;
}

- (BOOL)scanHTMLColor:(DTColor * __autoreleasing*)color
{
	return [self scanHTMLColor:color HTMLName:NULL];
}

- (BOOL)scanHTMLColor:(DTColor * __autoreleasing*)color HTMLName:(NSString * __autoreleasing*)name
{
	NSUInteger indexBefore = [self scanLocation];
	
	NSString *colorName = nil;
	
	NSMutableCharacterSet *tokenEndSet = [[NSCharacterSet whitespaceAndNewlineCharacterSet] mutableCopy];
	[tokenEndSet addCharactersInString:@","];
	
	if ([self scanString:@"#" intoString:NULL])
	{
		self.scanLocation = indexBefore;
		
		[self scanUpToCharactersFromSet:tokenEndSet intoString:&colorName];
	}
	else if ([self scanString:@"rgb" intoString:NULL])
	{
		if ([self scanUpToString:@")" intoString:NULL])
		{
			self.scanLocation++;
			colorName = [[self string] substringWithRange:NSMakeRange(indexBefore, self.scanLocation - indexBefore)];
			
			colorName = [colorName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
		}
	}
	else
	{
		// could be a plain html color name
		[self scanCharactersFromSet:[NSCharacterSet alphanumericCharacterSet] intoString:&colorName];
	}
	
	DTColor *foundColor = nil;
	
	if (colorName)
	{
		foundColor = DTColorCreateWithHTMLName(colorName);
	}
	
	if (!foundColor)
	{
		self.scanLocation = indexBefore;
		return NO;
	}

	if (color)
	{
		*color = foundColor;
	}

	if (name) {
		*name = colorName;
	}

	return YES;
}

@end

File: Sources/DTCoreTextSwift/NSScanner+HTML.swift
//
//  NSScanner+HTML.swift
//  DTCoreText
//
//  Created by Oliver Drobnik on 1/12/11.
//  Copyright 2011 Drobnik.com. All rights reserved.
//

import Foundation

/// Extensions for `Scanner` to deal with HTML-specific parsing, primarily CSS-related things.
extension Scanner {

  /// Scans a single CSS attribute (e.g. `color: red;`) out of a CSS style declaration.
  ///
  /// - Returns: A tuple of `(name, value)` on success, or `nil` if no valid attribute starts
  ///   at the current scan position. The `value` is either a `String` (for zero or one token)
  ///   or `[String]` (for comma-separated lists such as `font-family`).
  public func scanCSSAttribute() -> (name: String, value: Any)? {
    let initialScanLocation = currentIndex

    let whiteCharacterSet = CharacterSet.whitespacesAndNewlines

    var nonWhiteCommaCharacterSet = CharacterSet.whitespacesAndNewlines
    nonWhiteCommaCharacterSet.formUnion(CharacterSet(charactersIn: ";,"))
    nonWhiteCommaCharacterSet.invert()

    let cssStyleAttributeNameCharacterSet = NSCharacterSet.dt_cssStyleAttributeNameCharacterSet

    // scan attribute name
    guard let attrName = scanCharacters(from: cssStyleAttributeNameCharacterSet) else {
      return nil
    }

    // skip whitespace
    _ = scanCharacters(from: whiteCharacterSet)

    // expect :
    guard scanString(":") != nil else {
      currentIndex = initialScanLocation
      return nil
    }

    // skip whitespace
    _ = scanCharacters(from: whiteCharacterSet)

    var results: [String] = []
    var nextIterationAddsNewEntry = true

    while !isAtEnd && scanString(";") == nil && scanString("'';") == nil
      && scanString("\"\";") == nil
    {
      // skip whitespace
      _ = scanCharacters(from: whiteCharacterSet)

      if let quoteStr = scanCharacters(from: NSCharacterSet.dt_quoteCharacterSet) {
        // attribute is quoted
        guard let quotedValue = scanUpToString(quoteStr) else {
          currentIndex = initialScanLocation
          return nil
        }

        if nextIterationAddsNewEntry {
          results.append(quotedValue)
          nextIterationAddsNewEntry = false
        } else if let last = results.last {
          results[results.count - 1] = "\(last) \(quoteStr)\(quotedValue)\(quoteStr)"
        }

        // skip ending quote
        _ = scanString(quoteStr)
      } else if let scanned = scanString("rgb(") {
        let rgbContent = scanUpToString(";") ?? ""
        let formattedRGBString = "\(scanned)\(rgbContent)"

        if nextIterationAddsNewEntry {
          results.append(formattedRGBString)
          nextIterationAddsNewEntry = false
        } else if let last = results.last {
          results[results.count - 1] = "\(last) \(formattedRGBString)"
        }
      } else if let scanned = scanString(",") {
        var isStringOnlyCSSProperty = false

        if attrName == "font"
          || attrName.range(of: "color") != nil
          || attrName.range(of: "shadow") != nil
          || attrName.range(of: "background") != nil
        {
          if let last = results.last {
            results[results.count - 1] = "\(last)\(scanned)"
          }
          isStringOnlyCSSProperty = true
        }

        if !isStringOnlyCSSProperty {
          nextIterationAddsNewEntry = true
        }
      } else if let vs = scanCharacters(from: nonWhiteCommaCharacterSet) {
        if !vs.isEmpty && vs != "," {
          if nextIterationAddsNewEntry {
            results.append(vs)
            nextIterationAddsNewEntry = false
          } else if let last = results.last {
            results[results.count - 1] = "\(last) \(vs)"
          }
        }
      }

      // skip whitespace
      _ = scanCharacters(from: whiteCharacterSet)
    }

    let value: Any
    switch results.count {
    case 0: value = ""
    case 1: value = results[0]
    default: value = results
    }
    return (attrName, value)
  }

  /// Scans a CSS `url(...)` construct.
  ///
  /// - Returns: The decoded URL string (with HTML entities replaced), or `nil` if no
  ///   `url(...)` construct starts at the current scan position. Returns an empty string
  ///   for an empty construct like `url("")`.
  @discardableResult
  public func scanCSSURL() -> String? {
    guard scanString("url(") != nil else {
      return nil
    }

    let quoteCharacterSet = NSCharacterSet.dt_quoteCharacterSet
    var attrValue: String?

    if let quoteStr = scanCharacters(from: quoteCharacterSet) {
      if quoteStr.count == 1 {
        attrValue = scanUpToString(quoteStr)
        _ = scanString(quoteStr)
      } else {
        // most likely e.g. href=""
        attrValue = ""
      }

      attrValue = attrValue?.stringByReplacingHTMLEntities()
    } else if let scanned = scanUpToString(")") {
      // non-quoted attribute, ends at )
      attrValue = scanned.stringByReplacingHTMLEntities()
    }

    return attrValue
  }

  /// Scans a typical HTML color (hex, `rgb(...)`, or a named color).
  ///
  /// - Returns: A tuple of `(color, name)` on success, or `nil` if no valid color starts
  ///   at the current scan position. The scanner is left unchanged on failure.
  public func scanHTMLColor() -> (color: DTColor, name: String)? {
    let indexBefore = currentIndex

    var colorName: String?

    var tokenEndSet = CharacterSet.whitespacesAndNewlines
    tokenEndSet.insert(charactersIn: ",")

    if scanString("#") != nil {
      currentIndex = indexBefore
      colorName = scanUpToCharacters(from: tokenEndSet)
    } else if scanString("rgb") != nil {
      if scanUpToString(")") != nil {
        let str = self.string
        let afterRGB = str.index(after: currentIndex)
        if afterRGB <= str.endIndex {
          currentIndex = afterRGB
        }
        colorName = String(str[indexBefore..<currentIndex])
          .trimmingCharacters(in: .whitespacesAndNewlines)
      }
    } else {
      // could be a plain html color name
      colorName = scanCharacters(from: .alphanumerics)
    }

    guard let name = colorName,
      let color = DTColorCreateWithHTMLName(name)
    else {
      currentIndex = indexBefore
      return nil
    }

    return (color, name)
  }
}

// MARK: - Private helper to call stringByReplacingHTMLEntities on String

extension String {
  fileprivate func stringByReplacingHTMLEntities() -> String {
    return (self as NSString).stringByReplacingHTMLEntities()
  }
}
